instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy

```diff
@@ -0,0 +1,126 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.armeria.v_0_99
+
+import com.linecorp.armeria.client.WebClient
+import com.linecorp.armeria.common.HttpMethod
+import com.linecorp.armeria.common.HttpRequest
+import com.linecorp.armeria.common.HttpResponse
+import com.linecorp.armeria.common.HttpStatus
+import com.linecorp.armeria.server.ServerBuilder
+import com.linecorp.armeria.testing.junit4.server.ServerRule
+import io.opentelemetry.auto.test.InstrumentationSpecification
+import io.opentelemetry.auto.test.utils.TraceUtils
+import io.opentelemetry.trace.attributes.SemanticAttributes
+import org.junit.ClassRule
+import spock.lang.Ignore
+import spock.lang.Shared
+
+import static io.opentelemetry.trace.Span.Kind.SERVER
+
+abstract class AbstractArmeriaServerTest extends InstrumentationSpecification {
+
+  abstract void configureServer(ServerBuilder sb)
+
+  @ClassRule
+  @Shared
+  ServerRule server = new ServerRule() {
+    @Override
+    protected void configure(ServerBuilder sb) throws Exception {
+      sb.service("/exact", { ctx, req -> HttpResponse.of(HttpStatus.OK) })
+      sb.service("/items/{itemsId}", { ctx, req -> HttpResponse.of(HttpStatus.OK) })
+      sb.service("/httperror", { ctx, req -> HttpResponse.of(HttpStatus.NOT_IMPLEMENTED) })
+      sb.service("/exception", { ctx, req -> throw new IllegalStateException("illegal") })
+
+      configureServer(sb)
+    }
+  }
+
+  def client = WebClient.of(server.httpUri())
+
+  def "HTTP #method #path"() {
+    when:
+    def response = client.execute(HttpRequest.of(method, path)).aggregate().join()
+
+    then:
+    response.status().code() == code
+    assertTraces(1) {
+      trace(0, 1) {
+        span(0) {
+          operationName(spanName)
+          spanKind SERVER
+          errored code != 200
+          attributes {
+            "${SemanticAttributes.NET_PEER_IP.key()}" "127.0.0.1"
+            "${SemanticAttributes.NET_PEER_PORT.key()}" Long
+            "${SemanticAttributes.HTTP_URL.key()}" "${server.httpUri()}${path}"
+            "${SemanticAttributes.HTTP_METHOD.key()}" method.name()
+            "${SemanticAttributes.HTTP_STATUS_CODE.key()}" code
+
+            if (path == "/exception") {
+              "error.type" String
+              "error.msg" String
+              "error.stack" String
+            }
+          }
+        }
+      }
+    }
+
+    where:
+    path          | spanName     | method          | code
+    "/exact"      | "/exact"     | HttpMethod.GET  | 200
+    // TODO(anuraaga): Seems to be an Armeria bug not to have :objectId here
+    "/items/1234" | "/items/:"   | HttpMethod.POST | 200
```

Comment:
IIRC it's working as expected, although not very nice: https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/ParameterizedPathMapping.java#L64-L70

Armeria's `PathMapping` probably needs to provide a dedicated getter method for distributed tracing spans.

Reply:
Having a method like that to give a human readable representation of the route would be nice.

Reply:
Created an issue for solving this better: https://github.com/line/armeria/issues/2932